### DISCUSSION TITLE
modbus rtu problem

### DISCUSSION THREAD

**Discussion by defela** [aldas/modbus-tcp-client#121] [Q&A]
hello aldas, first of all, thank you, but I have a problem, I connected as modbus rtu, but I get the following package error, I make 10-15 queries, it gives as an array, then it is the same again, package parsing error, can you help?

```
RTU Binary received (in hex): 01030401a100d4aa �
An exception occurred Packet crc (\xd4aa) does not match calculated crc (\x6caa)! #0 C:\xampp\htdocs\fc4.php(35): ModbusTcpClient\Packet\RtuConverter::fromRtu('\x01\x03\x04\x01\xA1\x00\xD4\xAA') #1 {main}
```

```php
<?php

use ModbusTcpClient\Network\BinaryStreamConnection;
use ModbusTcpClient\Packet\ModbusFunction\ReadHoldingRegistersRequest;
use ModbusTcpClient\Packet\RtuConverter;
use ModbusTcpClient\Packet\ModbusFunction\ReadHoldingRegistersResponse;

require __DIR__ . '/vendor/autoload.php';
//require __DIR__ . '/logger.php';
$connection = BinaryStreamConnection::getBuilder()
    ->setPort(1001)
    ->setHost('10.20.0.74')
    ->setReadTimeoutSec(3) // increase read timeout to 3 seconds
    ->setIsCompleteCallback(function ($binaryData, $streamIndex) {
        // Do not check for complete TCP packet structure. Default implementation works only for Modbus TCP.
        // Modbus TCP has 7 byte header and this function checks for it and whole packet to be complete. RTU does
        // not have that.
        // Read about differences here: https://www.simplymodbus.ca/TCP.htm
        return true;
    })
   // ->setLogger(new EchoLogger())
    ->build();

$startAddress = 4;
$quantity = 2;
$slaveId = 1; // RTU packet slave id equivalent is Modbus TCP unitId

$tcpPacket = new ReadHoldingRegistersRequest($startAddress, $quantity, $slaveId);
$rtuPacket = RtuConverter::toRtu($tcpPacket);

try {
    $binaryData = $connection->connect()->sendAndReceive($rtuPacket);
    echo 'RTU Binary received (in hex):   ' . unpack('H*', $binaryData)[1] . PHP_EOL;

    $response = RtuConverter::fromRtu($binaryData);
    echo 'Parsed packet (in hex):     ' . $response->toHex() . PHP_EOL;
    echo 'Data parsed from packet (bytes):' . PHP_EOL;
    print_r($response->getData());

} catch (Exception $exception) {
    echo 'An exception occurred' . PHP_EOL;
    echo $exception->getMessage() . PHP_EOL;
    echo $exception->getTraceAsString() . PHP_EOL;
} finally {
    $connection->close();
}

```

**Comment by defela** [comment]
```php
<?php

use ModbusTcpClient\Network\BinaryStreamConnection;
use ModbusTcpClient\Packet\ModbusFunction\ReadHoldingRegistersRequest;
use ModbusTcpClient\Packet\RtuConverter;
use ModbusTcpClient\Packet\ModbusFunction\ReadHoldingRegistersResponse;

require __DIR__ . '/vendor/autoload.php';
//require __DIR__ . '/logger.php';
$connection = BinaryStreamConnection::getBuilder()
    ->setPort(1001)
    ->setHost('10.20.0.74')
    ->setReadTimeoutSec(3) // increase read timeout to 3 seconds
    ->setIsCompleteCallback(function ($binaryData, $streamIndex) {
        // Do not check for complete TCP packet structure. Default implementation works only for Modbus TCP.
        // Modbus TCP has 7 byte header and this function checks for it and whole packet to be complete. RTU does
        // not have that.
        // Read about differences here: https://www.simplymodbus.ca/TCP.htm
        return true;
    })
   // ->setLogger(new EchoLogger())
    ->build();

$startAddress = 4;
$quantity = 2;
$slaveId = 1; // RTU packet slave id equivalent is Modbus TCP unitId

$tcpPacket = new ReadHoldingRegistersRequest($startAddress, $quantity, $slaveId);
$rtuPacket = RtuConverter::toRtu($tcpPacket);

try {
    $binaryData = $connection->connect()->sendAndReceive($rtuPacket);
    echo 'RTU Binary received (in hex):   ' . unpack('H*', $binaryData)[1] . PHP_EOL;

    $response = RtuConverter::fromRtu($binaryData);
    echo 'Parsed packet (in hex):     ' . $response->toHex() . PHP_EOL;
    echo 'Data parsed from packet (bytes):' . PHP_EOL;
    print_r($response->getData());

} catch (Exception $exception) {
    echo 'An exception occurred' . PHP_EOL;
    echo $exception->getMessage() . PHP_EOL;
    echo $exception->getTraceAsString() . PHP_EOL;
} finally {
    $connection->close();
}
```

**Reply by defela** [reply]
```
RTU Binary received (in hex): 01030401a300d34a70 ��
Parsed packet (in hex): 44010000000701030401a300d3 
Data parsed from packet (bytes): Array ( [0] => 1 [1] => 163 [2] => 0 [3] => 211 )
```
This screen comes up after 10-15 tries.

**Reply by aldas** [reply]
@defela could you try https://www.modbusdriver.com/modpoll.html and see if that tool works or it occasionally gives also errors. I am just guessing but it could be some interference or faulty cables that bit-flips sometimes some bits and make CRC check fail.

See this block https://github.com/aldas/modbus-tcp-client/wiki/Debugging-guide#2-try-another-modbus-client but modify this command to use RTU 

Some explanation about returned data:

For example response `01030401a300d34a70` is
```
01 = unit id
03 = function code
04 = result has 4 bytes (2 registers)
01a300d3 = 2 registers of data (2*2 bytes)
4a70 = crc (crc is 2 bytes)
```

For example problematic response `01030401a100d4aa`

```
01 = unit id
03 = function code
04 = result has 4 bytes (2 registers)
01a100d4 = 2 registers of data (2*2 bytes)
aa = crc (SHOULD BE 2 bytes)
```

I do not know why but it seems that sometimes we are missing some bytes

**Reply by aldas** [reply]
@defela could you try this - this is improved function that allows us to properly determine if we have received enough bytes to consider data to be full/complete packet and ready to be return/parsed. There are very rare cases when we read only partial packet from connection and in that case we should retry reading.

```php
    ->setIsCompleteCallback(function ($binaryData, $streamIndex) {
        // minimal RTU packet length is 5 bytes (1 byte unit id + 1 byte function code + 1 byte of error code or byte length + 2 bytes for CRC)
        $length = strlen($binaryData);
        if ($length < 5) {
            return false;
        }
        if ((ord($binaryData[1]) & ErrorResponse::EXCEPTION_BITMASK) > 0) { // seems to be error response
            return true;
        }
        // if it is not error response then 3rd byte contains data length in bytes

        // leading 3 bytes are = unit id + function code + data length in bytes
        // next is N bytes of data that should match 3rd byte value
        // and 2 bytes for CRC
        // so adding these number is what complete packet would be
        $expectedLength = 3 + ord($binaryData[2]) + 2;

        if ($length > $expectedLength) {
            throw new IOException('packet length more bytes than expected');
        }
        return $length === $expectedLength;
    })
```

nb: I have not tested this

**Reply by defela** [reply]
```php
<?php

use ModbusTcpClient\Network\BinaryStreamConnection;
use ModbusTcpClient\Packet\ModbusFunction\ReadHoldingRegistersRequest;
use ModbusTcpClient\Packet\RtuConverter;
use ModbusTcpClient\Packet\ModbusFunction\ReadHoldingRegistersResponse;

require __DIR__ . '/vendor/autoload.php';
//require __DIR__ . '/logger.php';
$connection = BinaryStreamConnection::getBuilder()
    ->setPort(1001)
    ->setHost('10.20.0.74')
    ->setReadTimeoutSec(3) // increase read timeout to 3 seconds
    ->setIsCompleteCallback(function ($binaryData, $streamIndex) {
        // minimal RTU packet length is 5 bytes (1 byte unit id + 1 byte function code + 1 byte of error code or byte length + 2 bytes for CRC)
        $length = strlen($binaryData);
        if ($length < 5) {
            return false;
        }
        if ((ord($binaryData[1]) & ErrorResponse::EXCEPTION_BITMASK) > 0) { // seems to be error response
            return true;
        }
        // if it is not error response then 3rd byte contains data length in bytes

        // leading 3 bytes are = unit id + function code + data length in bytes
        // next is N bytes of data that should match 3rd byte value
        // and 2 bytes for CRC
        // so adding these number is what complete packet would be
        $expectedLength = 3 + ord($binaryData[2]) + 2;

        if ($length > $expectedLength) {
            throw new IOException('packet length more bytes than expected');
        }
        return $length === $expectedLength;
    })
   // ->setLogger(new EchoLogger())
    ->build();

$startAddress = 4;
$quantity = 2;
$slaveId = 1; // RTU packet slave id equivalent is Modbus TCP unitId

$tcpPacket = new ReadHoldingRegistersRequest($startAddress, $quantity, $slaveId);
$rtuPacket = RtuConverter::toRtu($tcpPacket);

try {
    $binaryData = $connection->connect()->sendAndReceive($rtuPacket);
    echo 'RTU Binary received (in hex):   ' . unpack('H*', $binaryData)[1] . PHP_EOL;

    $response = RtuConverter::fromRtu($binaryData);
    echo 'Parsed packet (in hex):     ' . $response->toHex() . PHP_EOL;
    echo 'Data parsed from packet (bytes):' . PHP_EOL;
    print_r($response->getData());

} catch (Exception $exception) {
    echo 'An exception occurred' . PHP_EOL;
    echo $exception->getMessage() . PHP_EOL;
    echo $exception->getTraceAsString() . PHP_EOL;
} finally {
    $connection->close();
}
```

**Reply by defela** [reply]
`Fatal error: Uncaught Error: Class "ErrorResponse" not found in C:\xampp\htdocs\fc4.php:20 Stack trace: #0 C:\xampp\htdocs\vendor\aldas\modbus-tcp-client\src\Network\StreamHandler.php(78): {closure}('\x01\x03\x04\x01\x87\x00\xDD\x8B\xBF', 0) #1 C:\xampp\htdocs\vendor\aldas\modbus-tcp-client\src\Network\BinaryStreamConnection.php(60): ModbusTcpClient\Network\BinaryStreamConnection->receiveFrom(Array, 3.0, NULL) #2 C:\xampp\htdocs\vendor\aldas\modbus-tcp-client\src\Network\BinaryStreamConnection.php(83): ModbusTcpClient\Network\BinaryStreamConnection->receive() #3 C:\xampp\htdocs\fc4.php(47): ModbusTcpClient\Network\BinaryStreamConnection->sendAndReceive('\x01\x03\x00\x04\x00\x02\x85\xCA') #4 {main} thrown in C:\xampp\htdocs\fc4.php on line 20`

**Reply by defela** [reply]
modbuspool with modbus/rtu/ascii over read success but php no

**Reply by aldas** [reply]
you need import `use ModbusTcpClient\Packet\ErrorResponse;`

**Reply by defela** [reply]
aldas thank you very much but now there is such a problem modbus pool read values ​​php read values ​​slave id 1 registry 1 quan 10 why is that

**Reply by defela** [reply]
![rtu](https://user-images.githubusercontent.com/51213063/195975036-35c18375-6139-4708-9b5f-c2bc3e169179.PNG)

**Reply by defela** [reply]
![rtu2](https://user-images.githubusercontent.com/51213063/195975152-8c36792f-8eb0-49f7-9058-f1af1972e40f.PNG)

slave 1 register 4 quan 2 but value problem

**Reply by aldas** [reply]
`print_r($response->getData());`  prints received data as bytes. For each returned register there is 2 bytes.  For example 4 bytes of data

```
data[0] = 1 /// in hex 0x01
data[1] = 140 // in hex 0x8C
data[2] = 0 // in hex 0x0
data[3] = 225 // in hex
```

can be interpreted as 2 register where 
* first register data is  [0x1, 0x8c]  so 0x018c  which is in decimal 396   
* second register data is [0x00, 0xE1] so 0x00E1 which is in decimal 225

It would be probably easier if you would use https://github.com/aldas/modbus-tcp-client/blob/master/examples/example_parallel_requests.php  `ReadRegistersBuilder::newReadHoldingRegisters` to build your requests as you can use data type methods for returned data.

something like that
```php
<?php

use ModbusTcpClient\Composer\Read\ReadRegistersBuilder;
use ModbusTcpClient\Network\BinaryStreamConnection;
use ModbusTcpClient\Network\IOException;
use ModbusTcpClient\Packet\ErrorResponse;
use ModbusTcpClient\Packet\RtuConverter;

require __DIR__ . '/../vendor/autoload.php';
require __DIR__ . '/logger.php';

$connection = BinaryStreamConnection::getBuilder()
    ->setPort(1001)
    ->setHost('10.20.0.74')
    ->setReadTimeoutSec(3) // increase read timeout to 3 seconds
    ->setIsCompleteCallback(function ($binaryData, $streamIndex) {
        // minimal RTU packet length is 5 bytes (1 byte unit id + 1 byte function code + 1 byte of error code or byte length + 2 bytes for CRC)
        $length = strlen($binaryData);
        if ($length < 5) {
            return false;
        }
        if ((ord($binaryData[1]) & ErrorResponse::EXCEPTION_BITMASK) > 0) { // seems to be error response
            return true;
        }
        // if it is not error response then 3rd byte contains data length in bytes

        // leading 3 bytes are = unit id + function code + data length in bytes
        // next is N bytes of data that should match 3rd byte value
        // and 2 bytes for CRC
        // so adding these number is what complete packet would be
        $expectedLength = 3 + ord($binaryData[2]) + 2;

        if ($length > $expectedLength) {
            throw new IOException('packet length more bytes than expected');
        }
        return $length === $expectedLength;
    })
    ->setLogger(new EchoLogger())
    ->build();

$address = 'tcp://127.0.0.1:5022';
$unitID = 0;
$fc3requests = ReadRegistersBuilder::newReadHoldingRegisters($address, $unitID)
    ->int16(4, 'address4_value') // or uint16() - whatever data type that value is in that register
    ->int16(5, 'address5_value')
    ->build(); // returns array of ReadHoldingRegistersRequest requests

try {
    foreach ($fc3requests as $request) {
        echo 'Packet to be sent (in hex): ' . $request->toHex() . PHP_EOL;
        $rtuPacket = RtuConverter::toRtu($request->getRequest());

        $binaryData = $connection->connect()->sendAndReceive($rtuPacket);
        echo 'RTU Binary received (in hex):   ' . unpack('H*', $binaryData)[1] . PHP_EOL;

        $tcpResponsePacket = RtuConverter::fromRtu($binaryData);
        
        echo 'Data parsed from packet (bytes):' . PHP_EOL;
        $result = $request->parse($tcpResponsePacket);
        print_r($result);
    }
} catch (Exception $exception) {
    echo 'An exception occurred' . PHP_EOL;
    echo $exception->getMessage() . PHP_EOL;
    echo $exception->getTraceAsString() . PHP_EOL;
} finally {
    $connection->close();
}
```

NB: I have not tested this. if you are missing requires/imports add them etc

**Reply by defela** [reply]
aldas sorry but i couldn't do it this is my code is it possible for you to compile and send it or can you help me if i give anydesk

**Reply by defela** [reply]
```php
<?php

use ModbusTcpClient\Network\BinaryStreamConnection;
use ModbusTcpClient\Packet\ModbusFunction\ReadHoldingRegistersRequest;
use ModbusTcpClient\Packet\RtuConverter;
use ModbusTcpClient\Packet\ModbusFunction\ReadHoldingRegistersResponse;
use ModbusTcpClient\Packet\ResponseFactory;
use ModbusTcpClient\Packet\ErrorResponse;

require __DIR__ . '/vendor/autoload.php';
//require __DIR__ . '/logger.php';
$connection = BinaryStreamConnection::getBuilder()
    ->setPort(1001)
    ->setHost('10.20.0.74')
    ->setReadTimeoutSec(5) // increase read timeout to 3 seconds
    ->setIsCompleteCallback(function ($binaryData, $streamIndex) {
        // minimal RTU packet length is 5 bytes (1 byte unit id + 1 byte function code + 1 byte of error code or byte length + 2 bytes for CRC)
        $length = strlen($binaryData);
        if ($length < 5) {
            return false;
        }
        if ((ord($binaryData[1]) & ErrorResponse::EXCEPTION_BITMASK) > 0) { // seems to be error response
            return true;
        }
        // if it is not error response then 3rd byte contains data length in bytes

        // leading 3 bytes are = unit id + function code + data length in bytes
        // next is N bytes of data that should match 3rd byte value
        // and 2 bytes for CRC
        // so adding these number is what complete packet would be
        $expectedLength = 3 + ord($binaryData[2]) + 2;

        if ($length > $expectedLength) {
            throw new IOException('packet length more bytes than expected');
        }
        return $length === $expectedLength;
    })
   // ->setLogger(new EchoLogger())
    ->build();

$startAddress = 4;
$quantity = 2;
$slaveId = 1; // RTU packet slave id equivalent is Modbus TCP unitId

$tcpPacket = new ReadHoldingRegistersRequest($startAddress, $quantity, $slaveId);
$rtuPacket = RtuConverter::toRtu($tcpPacket);

try {
    $binaryData = $connection->connect()->sendAndReceive($rtuPacket);
    echo 'RTU Binary received (in hex):   ' . unpack('H*', $binaryData)[1] . PHP_EOL;

    $response = RtuConverter::fromRtu($binaryData);
    echo 'Parsed packet (in hex):     ' . $response->toHex() . PHP_EOL;
    echo 'Data parsed from packet (bytes):' . PHP_EOL;
    print_r($response->getData());

} catch (Exception $exception) {
    echo 'An exception occurred' . PHP_EOL;
    echo $exception->getMessage() . PHP_EOL;
    echo $exception->getTraceAsString() . PHP_EOL;
} finally {
    $connection->close();
}
```

**Reply by defela** [reply]
anydesk id : 337104046

**Reply by defela** [reply]
I would be very happy

**Reply by aldas** [reply]
Hi, I have release new version `3.1.0` with some RTU related improvements. And added new example with RTU over TCP https://github.com/aldas/modbus-tcp-client/blob/master/examples/rtu_over_tcp_with_higherlevel_api.php which should work. At least when I tested with https://www.modbusdriver.com/diagslave.html it worked

**Comment by defela** [comment]
aldas thanks a lot i read it but do you have any idea about this problem multiple connections at the same time

**Reply by defela** [reply]
`<?php

use ModbusTcpClient\Composer\Read\ReadRegistersBuilder;
use ModbusTcpClient\Composer\Read\Register\ReadRegisterRequest;
use ModbusTcpClient\Network\BinaryStreamConnection;
use ModbusTcpClient\Packet\RtuConverter;
use ModbusTcpClient\Utils\Packet;

require 'C:\xampp\htdocs\vendor\autoload.php';
//require 'logger.php';

// Modbus server simulator https://www.modbusdriver.com/diagslave.html
// Start simulator with `./diagslave -m enc -a 1 -p 5020`


// $PORT     = $_GET["PORT"];   //1001
// $IP       = $_GET["IP"];    //10.20.074
// $UNITID   = $_GET["UNITID"];  //1
// $REG1     = $_GET["REG1"];  // 4
// $REG1_VALUE = $_GET["REG1_VALUE"];  //ISI
// $REG2     = $_GET["REG2"]; //5  
// $REG2_VALUE = $_GET["REG2_VALUE"];  //NEM

function getRTU($PORT,$IP,$UNITID,$REG1,$REG1_VALUE,$REG2,$REG2_VALUE) { 



$connection = BinaryStreamConnection::getBuilder()
    ->setPort($PORT)
    ->setHost($IP)
    ->setReadTimeoutSec(0.5)
    ->setIsCompleteCallback(function ($binaryData, $streamIndex) {
        return Packet::isCompleteLengthRTU($binaryData);
    })
    //->setLogger(new EchoLogger())
    ->build();

$unitID = $UNITID; // RTU packet slave id equivalent is Modbus TCP unitId
$fc3requests = ReadRegistersBuilder::newReadHoldingRegisters('10.20.0.74', $unitID) // uri/address does not matter because we use $connection
    ->uint16($REG1, $REG1_VALUE)
    ->uint16($REG2, $REG2_VALUE)
   
    // See `ReadRegistersBuilder.php` for available data type methods
    ->build(); // returns array of ReadHoldingRegistersRequest requests

     

    try {
    /** @var $request ReadRegisterRequest */
    foreach ($fc3requests as $request) {
        $rtuPacket = RtuConverter::toRtu($request->getRequest());

        $binaryData = $connection->connect()->sendAndReceive($rtuPacket);
       
        $tcpResponsePacket = RtuConverter::fromRtu($binaryData);

         $result = $request->parse($tcpResponsePacket);
                 
    }
} catch (Exception $exception) {
    echo 'An exception occurred' . PHP_EOL;
    echo $exception->getMessage() . PHP_EOL;
    echo $exception->getTraceAsString() . PHP_EOL;
} finally {
    $connection->close();
}

  return $result;

}
  
 

//echo $r0["ISI"];`

**Reply by defela** [reply]
`<?php 


   include("rtu_over_tcp_with_higherlevel_api.php");
   $DATA =  getRTU('1001','10.20.0.74','1','4','ISI','5','NEM'); //&REG1=4&REG1_VALUE=ISI&REG2=5&REG2_VALUE=NEM');
   $DATA2 =  getRTU('1001','10.20.0.74','2','4','ISI','5','NEM'); //&REG1=4&REG1_VALUE=ISI&REG2=5&REG2_VALUE=NEM');
   
   
   echo "<pre>";
   print_r($DATA);
   print_r($DATA2);

   echo "</pre>";
   echo "<pre>";


?>`

**Reply by defela** [reply]
Errors occur frequently when I read at the same time

**Reply by defela** [reply]
Errors occur frequently when I read at the same time

**Reply by defela** [reply]
`
Notice: fwrite(): Send of 8 bytes failed with errno=10054 Varolan bir bağlantı uzaktaki bir ana bilgisayar tarafından zorla kapatıldı in C:\xampp\htdocs\vendor\aldas\modbus-tcp-client\src\Network\BinaryStreamConnection.php on line 74
An exception occurred fread error during receiveFrom #0 C:\xampp\htdocs\vendor\aldas\modbus-tcp-client\src\Network\BinaryStreamConnection.php(60): ModbusTcpClient\Network\BinaryStreamConnection->receiveFrom(Array, 0.5, NULL) #1 C:\xampp\htdocs\vendor\aldas\modbus-tcp-client\src\Network\BinaryStreamConnection.php(83): ModbusTcpClient\Network\BinaryStreamConnection->receive() #2 C:\xampp\htdocs\vendor\modbus-tcp-client-master\examples\rtu_over_tcp_with_higherlevel_api.php(53): ModbusTcpClient\Network\BinaryStreamConnection->sendAndReceive('\x02\x03\x00\x04\x00\x02\x85\xF9') #3 C:\xampp\htdocs\vendor\modbus-tcp-client-master\examples\alp.php(6): getRTU('1001', '10.20.0.74', '2', '4', 'ISI', '5', 'NEM') #4 {main}
Warning: Undefined variable $result in C:\xampp\htdocs\vendor\modbus-tcp-client-master\examples\rtu_over_tcp_with_higherlevel_api.php on line 68
An exception occurred Unable to create client socket to tcp://10.20.0.74:1001: Bağlanılan uygun olarak belli bir süre içinde yanıt vermediğinden veya kurulan bağlantı bağlanılan ana bilgisayar yanıt vermediğinden bir bağlantı kurulamadı #0 C:\xampp\htdocs\vendor\aldas\modbus-tcp-client\src\Network\BinaryStreamConnectionBuilder.php(33): ModbusTcpClient\Network\InternetDomainStreamCreator->createStream(Object(ModbusTcpClient\Network\BinaryStreamConnection)) #1 C:\xampp\htdocs\vendor\aldas\modbus-tcp-client\src\Network\BinaryStreamConnection.php(41): ModbusTcpClient\Network\BinaryStreamConnectionBuilder->ModbusTcpClient\Network\{closure}(Object(ModbusTcpClient\Network\BinaryStreamConnection)) #2 C:\xampp\htdocs\vendor\modbus-tcp-client-master\examples\rtu_over_tcp_with_higherlevel_api.php(53): ModbusTcpClient\Network\BinaryStreamConnection->connect() #3 C:\xampp\htdocs\vendor\modbus-tcp-client-master\examples\alp.php(7): getRTU('1001', '10.20.0.74', '3', '4', 'ISI', '5', 'NEM') #4 {main}
Warning: Undefined variable $result in C:\xampp\htdocs\vendor\modbus-tcp-client-master\examples\rtu_over_tcp_with_higherlevel_api.php on line 68
An exception occurred fread error during receiveFrom #0 C:\xampp\htdocs\vendor\aldas\modbus-tcp-client\src\Network\BinaryStreamConnection.php(60): ModbusTcpClient\Network\BinaryStreamConnection->receiveFrom(Array, 0.5, NULL) #1 C:\xampp\htdocs\vendor\aldas\modbus-tcp-client\src\Network\BinaryStreamConnection.php(83): ModbusTcpClient\Network\BinaryStreamConnection->receive() #2 C:\xampp\htdocs\vendor\modbus-tcp-client-master\examples\rtu_over_tcp_with_higherlevel_api.php(53): ModbusTcpClient\Network\BinaryStreamConnection->sendAndReceive('\x05\x03\x00\x04\x00\x02\x84N') #3 C:\xampp\htdocs\vendor\modbus-tcp-client-master\examples\alp.php(9): getRTU('1001', '10.20.0.74', '5', '4', 'ISI', '5', 'NEM') #4 {main}
Warning: Undefined variable $result in C:\xampp\htdocs\vendor\modbus-tcp-client-master\examples\rtu_over_tcp_with_higherlevel_api.php on line 68
Array
(
    [ISI] => 473
    [NEM] => 224
)
Array
(
    [ISI] => 506
    [NEM] => 210
)`

**Reply by aldas** [reply]
> errno=10054 An existing connection was forcibly closed by the remote host.

could mean that your Modbus gateway has some kind of limit for connections or something. I you can try slowing down/reducing the amount of requests per second.

**Reply by defela** [reply]
There is no problem in reading a single value, but for example, when 2 connections are made at the same time, unit1 unit2 falls into error. Trying 4-5 times, it does not give a problem

**Reply by aldas** [reply]
as I said - you probably run into limitations of your Modbus serve/gateway software/hardware limits. This gateway probably uses serial or something like that to internally to talk with devices and is limited to single request concurrently.

**Reply by defela** [reply]
thank you very much

**Reply by defela** [reply]
aldas hello unfortunately, modbus rtu creates problems in more than one query, can we read from serial ports on windows machine instead? serial ports are connected to windows machine as com1 com2

**Reply by defela** [reply]
`<?php

use ModbusTcpClient\Packet\ModbusFunction\ReadInputRegistersRequest;
use ModbusTcpClient\Packet\ModbusFunction\ReadInputRegistersResponse;
use ModbusTcpClient\Packet\RtuConverter;

require __DIR__ . '/../vendor/autoload.php';

if (stripos(PHP_OS, 'WIN') === 0) {
    echo 'This example can not be run in Windows!' . PHP_EOL;
    exit(0);
}

$startAddress = 1;
$quantity = 2;
$slaveId = 1; // RTU packet slave id equivalent is Modbus TCP unitId

$tcpPacket = new ReadInputRegistersRequest($startAddress, $quantity, $slaveId);
$rtuPacket = RtuConverter::toRtu($tcpPacket);

try {
    $device = '/dev/ttyUSB0';
    $fd = fopen($device, 'w+b');

    $sttyModes = implode(' ', [
        'cs8', // enable character size 8 bits
        '9600', // enable baud rate 9600
        '-icanon', // disable enable special characters: erase, kill, werase, rprnt
        'min 0', // with -icanon, set N characters minimum for a completed read
        'ignbrk', // enable ignore break characters
        '-brkint', // disable breaks cause an interrupt signal
        '-icrnl', // disable translate carriage return to newline
        '-imaxbel', // disable beep and do not flush a full input buffer on a character
        '-opost', // disable postprocess output
        '-onlcr', // disable translate newline to carriage return-newline
        '-isig', // disable interrupt, quit, and suspend special characters
        '-iexten', // disable non-POSIX special characters
        '-echo', // disable echo input characters
        '-echoe', // disable echo erase characters as backspace-space-backspace
        '-echok', // disable echo a newline after a kill character
        '-echoctl', // disable same as [-]ctlecho
        '-echoke', // disable kill all line by obeying the echoprt and echoe settings
        '-noflsh', // disable flushing after interrupt and quit special characters
        '-ixon', // disable XON/XOFF flow control
        '-crtscts', // disable RTS/CTS handshaking
    ]);
    $sttyResult = exec("stty -F ${device} ${sttyModes}");
    if ($sttyResult === false) {
        echo 'stty command failed' . PHP_EOL;
        exit(1);
    }

    echo 'RTU Binary to sent (in hex):   ' . unpack('H*', $rtuPacket)[1] . PHP_EOL;
    fwrite($fd, $rtuPacket);
    fflush($fd);

    $binaryData = '';
    $start = microtime(true);

    do {
        // give sensor (5ms) some time to respond. SHT20 modbus minimal response time seems to be 20ms and more
        usleep(5000);
        $binaryData = fread($fd, 255);
    } while ($binaryData === '');

    $end = (microtime(true) - $start) * 1000;
    echo 'Response in: ' . $end . ' ms' . PHP_EOL;

    echo 'RTU Binary received (in hex):   ' . unpack('H*', $binaryData)[1] . PHP_EOL;

    /** @var ReadInputRegistersResponse $response */
    $response = RtuConverter::fromRtu($binaryData)->withStartAddress($startAddress);
    echo 'Parsed packet (in hex):     ' . $response->toHex() . PHP_EOL;

    echo PHP_EOL;
    echo 'Temperature: ' . ($response->getWordAt(1)->getInt16() / 10) . PHP_EOL;
    echo 'Humidity: ' . ($response->getWordAt(2)->getInt16() / 10) . PHP_EOL;

} catch (Exception $exception) {
    echo 'An exception occurred' . PHP_EOL;
    echo $exception->getMessage() . PHP_EOL;
    echo $exception->getTraceAsString() . PHP_EOL;
} finally {
    fclose($fd);
}`

**Reply by defela** [reply]
Here you said that if it is not a windows machine, how should I connect to the coms of the windows machine?

**Reply by aldas** [reply]
Maybe https://stackoverflow.com/a/1074273/2514290 ? I really no not know as I use Windows only for gaming and these days it is quite rare occasion.

```php
$output = exec("mode COM2: BAUD=115200 PARITY=N data=8 stop=1 XON=off TO=on dtr=off odsr=off octs=off rts=on idsr=off");
$fp = fopen("COM2", "r+");
```
